PhysicsTools/PatAlgos/plugins/JetCorrFactorsProducer.cc

```diff
@@ -232,12 +232,9 @@ JetCorrFactorsProducer::produce(edm::Event& event, const edm::EventSetup& setup)
 					    << "the sequence.";
     }
     for(unsigned int idx=0; idx<corrLevel->second.size(); ++idx){
-      bool flavorDependent=false;
       std::vector<float> factors;
-      if(flavorDependent ||
-	 corrLevel->second[idx].find("L5Flavor")!=std::string::npos ||
+      if(corrLevel->second[idx].find("L5Flavor")!=std::string::npos ||
```

Comment:
Copying discussion and Sal's answer to this thread:

>    Hi Sal,
>    Could you please clarify "This looks fine." Which of the following options is best:
    > 1. Maintain the original behavior, which is that L5Flavor and L7Parton
    > will be treated as flavor-dependent, and L6UE and any others will not.
    >
    > 2. Fix the code to match what it seems to intend, which is that as soon
    > as L5Flavor or L7Parton are encountered in the list, all the following
    > corrections will be flavor-dependent. If we assume the corrections are
    > always in numerical order, that would mean L6UE would be treated as
    > flavor-dependent.

 
Yes, sorry, I misunderstood, I forgot about the "L6UE" correction. This is completely deprecated so these would both give the same results in practice. The comment also neglects it, so that is probably where the confusion lies. 

If there are modest speed improvements by removing the check, it is safe to do so from my perspective. 

On Sat, Sep 24, 2016 at 9:25 PM, Slava Krutelyov <slava77@fnal.gov> wrote:

    On 9/24/16 11:33 AM, Salvatore Rappoccio wrote:
    > Hi, Guys,
    >
    > Sorry, folks, I don't always catch these comments inside git threads, so
    > thanks to Slava for the email reminder. This looks fine. We haven't used
    > these flavor dependent corrections in a long time, but perhaps we can
    > come up with one in the future.

    Hi Sal,

    Could you please be more explicit on the code change implied.


    I was staring at this code for a bit and it looks rather contrite.
    It operates with a map<flavor, vector<string (describes a corrector?)> >
    It looks like there is some implicit order assumed between vectors
    corresponding to different flavor keys.
    It looks like it has to be the same, else it doesn't make sense to check
    all other flavor keys at the same index.

    If the strings in this vector are completely independent, then the
    current code behavior is OK,
    and the "bool flavorDependent" variable is completely redundant.

 
The different corrections are not completely independent and do indeed have a specific order in which they must be applied. We have mitigated this to a large extent by combining "L2" and "L3" into one step. However, the pileup correction "L1" must be done before the nonlinearity corrections "L2L3". The flavor corrections "L5" are not used to my knowledge unless there are some private users doing it somewhere. The underlying event "L6UE" correction is totally deprecated. The shower-to-parton correction "L7" is also never used anymore unless some private users are doing it somewhere.

The order of the flavor keys is also set, it is "Gluon", "UDS", "charm", "bottom".